index.js

```diff
@@ -104,51 +77,279 @@ module.exports = {
+      if (jso.params.result === null || jso.params.result === undefined)
+        return this.configuration.errorHandler(new ErrorWithData("Received an `eth_subscription` request without a result.", jso));
+
+      var subscriptionCallback = this.internalState.subscriptions[jso.params.subscription];
+      if (subscriptionCallback)
+        subscriptionCallback(jso.params.result);
+    }.bind(this);
+
+    var responseHandler = function () {
+      if (typeof jso.id !== "number")
+        return this.configuration.errorHandler(new ErrorWithData("Received a message from the blockchain that didn't have a valid id.", jso));
+      var outstandingRequest = this.internalState.outstandingRequests[jso.id];
+      delete this.internalState.outstandingRequests[jso.id];
+      if (typeof outstandingRequest !== "object")
+        return this.configuration.errorHandler(new ErrorWithData("Unable to locate original request for blockchain response.", jso));
+
+      // FIXME: outstandingRequest.callback should be function(Error,object) not function(Error|object)
+      this.parse(jso, outstandingRequest.expectedReturnTypes, outstandingRequest.callback);
+    }.bind(this);
+
+    var errorHandler = function () {
+      // errors with IDs can go through the normal result process
+      if (jso.id !== null && jso.id !== undefined)
+        return responseHandler.bind(this)(jso);
+      this.configuration.errorHandler(new ErrorWithCodeAndData(jso.error.message, jso.error.code, jso.error.data));
+    }.bind(this);
+
+    // depending on the type of message it is (request, response, error, invalid) we will handle it differently
+    if (jso.method !== undefined) {
+      subscriptionHandler();
+    } else if (jso.result !== undefined) {
+      responseHandler();
+    } else if (jso.error !== undefined) {
+      errorHandler();
+    } else {
+      this.configuration.errorHandler(new ErrorWithData("Received an invalid JSON-RPC message.", jso));
+    }
   },
 
-  disableHostedNodeFallback: function () {
-    this.useHostedNodeFallback = false;
-    this.nodes.hosted = [];
-    if (this.wsUrl === HOSTED_WEBSOCKET) {
-      this.wsUrl = null;
+  /**
+   * Used internally.  Setup the initial subscription for new blocks.  Called on first connect, then never again.
+   * 
+   * @param {function():void} callback - called when the subscription is done being setup
+   */
+  setupBlockSubscription: function (callback) {
+    var handleSubscribeToNewHeadsResult = function(resultOrError) {
+      if (resultOrError instanceof Error || resultOrError.error) {
+        this.internalState.newBlockIntervalTimeoutId = setInterval(this.getBlockByNumber.bind(this, "latest", false, this.onNewBlock.bind(this, true)), this.BLOCK_POLL_INTERVAL);
+      } else {
+        this.internalState.subscriptions[resultOrError] = this.onNewBlock.bind(this, true);
+      }
+      callback();
+      callback = function () { };
```

Comment:
I added a comment (should have done this originally) to make it clear why this is happening.  The issue is that the same logic is used for processing both initial connect (which is when the callback makes sense) and reconnects (when calling the callback doesn't make sense).  By clearing the callback after calling it the first time (on initial connect) we can otherwise re-use all of the code for block subscription management on reconnect.